self-core-impl/src/main/java/com/selfxdsd/core/contracts/invoices/StoredPayment.java

```diff
@@ -0,0 +1,137 @@
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.selfxdsd.core.contracts.invoices;
+
+import com.selfxdsd.api.Invoice;
+import com.selfxdsd.api.Payment;
+
+import java.math.BigDecimal;
+import java.time.LocalDateTime;
+import java.util.Objects;
+
+/**
+ * A Payment stored in Self.
+ * @author Ali FELLAHI (fellahi.ali@gmail.com)
+ * @version $Id$
+ * @since 0.0.67
+ */
+public final class StoredPayment implements Payment {
+
+    /**
+     * Invoice to which this payment made.
+     */
+    private final Invoice invoice;
+
+    /**
+     * Successful Payment transaction id.
+     */
+    private final String transactionId;
+
+    /**
+     * Payment timestamp.
+     */
+    private final LocalDateTime paymentTime;
+
+    /**
+     * Payment amount.
+     */
+    private final BigDecimal value;
+
+    /**
+     * Payment status.
+     */
+    private final String status;
+
+    /**
+     * Payment failure reason.
+     */
+    private final String failReason;
+
+    /**
+     * Ctor.
+     * @param invoice The Invoice to which this payment made.
+     * @param transactionId Successful payment transaction id.
+     * @param paymentTime Payment time.
+     * @param value Payment amount.
+     * @param status Payment status (FAILED or SUCCESSFUL).
+     * @param failReason Payment failure reason.
+     */
+    public StoredPayment(final Invoice invoice,
+                         final String transactionId,
+                         final LocalDateTime paymentTime,
+                         final BigDecimal value,
+                         final String status,
+                         final String failReason
+    ) {
+        this.invoice = invoice;
+        this.transactionId = transactionId;
+        this.paymentTime = paymentTime;
+        this.value = value;
+        this.status = status;
+        this.failReason = failReason;
+    }
+
+    @Override
+    public Invoice invoice() {
+        return this.invoice;
+    }
+
+    @Override
+    public String transactionId() {
+        return this.transactionId;
+    }
+
+    @Override
+    public LocalDateTime paymentTime() {
+        return this.paymentTime;
+    }
+
+    @Override
+    public BigDecimal value() {
+        return this.value;
+    }
+
+    @Override
+    public String status() {
+        return this.status;
+    }
+
+    @Override
+    public String failReason() {
+        return this.failReason;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(
+            this.invoice.invoiceId(),
+            this.paymentTime()
+        );
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        return this == obj || (obj instanceof StoredPayment
+            && this.hashCode() == obj.hashCode());
```

Comment:
@fellahi-ali This method doesn't look right. The relation between equals and hashcode is:

1. If 2 objects are equal, then the hashcode must be identical.
BUT
2. If the hashcode is identical, it doesn't necessarily mean that the objects are equal. 

Maybe use the IDE to generate equals and hashcode based on paymentTime and transactionId?